integration-tests/tests/circuits.rs

```diff
@@ -35,21 +52,143 @@ const CIRCUITS_PARAMS: CircuitsParams = CircuitsParams {
     keccak_padding: None,
 };
 
-async fn test_evm_circuit_block(block_num: u64) {
-    log::info!("test evm circuit, block number: {}", block_num);
+async fn gen_inputs(
+    block_num: u64,
+) -> (
+    CircuitInputBuilder,
+    eth_types::Block<eth_types::Transaction>,
+) {
     let cli = get_client();
     let cli = BuilderClient::new(cli, CIRCUITS_PARAMS).await.unwrap();
-    let (builder, _) = cli.gen_inputs(block_num).await.unwrap();
+
+    cli.gen_inputs(block_num).await.unwrap()
+}
+
+fn test_run<ConcreteCircuit: Circuit<Fr>, R: RngCore>(
+    mut rng: R,
+    degree: u32,
+    circuit: ConcreteCircuit,
+    instance: Vec<Vec<Fr>>,
+) {
+    fn test_gen_proof<ConcreteCircuit: Circuit<Fr>, R: RngCore>(
+        rng: R,
+        circuit: ConcreteCircuit,
+        general_params: &ParamsKZG<Bn256>,
+        proving_key: &ProvingKey<G1Affine>,
+        mut transcript: Blake2bWrite<Vec<u8>, G1Affine, Challenge255<G1Affine>>,
+        instances: &[&[Fr]],
+    ) -> Vec<u8> {
+        create_proof::<
+            KZGCommitmentScheme<Bn256>,
+            ProverSHPLONK<'_, Bn256>,
+            Challenge255<G1Affine>,
+            R,
+            Blake2bWrite<Vec<u8>, G1Affine, Challenge255<G1Affine>>,
+            ConcreteCircuit,
+        >(
+            general_params,
+            proving_key,
+            &[circuit],
+            &[instances],
+            rng,
+            &mut transcript,
+        )
+        .expect("proof generation should not fail");
+
+        transcript.finalize()
+    }
+
+    fn test_verify(
+        general_params: &ParamsKZG<Bn256>,
+        verifier_params: &ParamsKZG<Bn256>,
+        verifying_key: &VerifyingKey<G1Affine>,
+        proof: &[u8],
+        instances: &[&[Fr]],
+    ) {
+        let mut verifier_transcript = Blake2bRead::<_, G1Affine, Challenge255<_>>::init(proof);
+        let strategy = SingleStrategy::new(general_params);
+
+        verify_proof::<
+            KZGCommitmentScheme<Bn256>,
+            VerifierSHPLONK<'_, Bn256>,
+            Challenge255<G1Affine>,
+            Blake2bRead<&[u8], G1Affine, Challenge255<G1Affine>>,
+            SingleStrategy<'_, Bn256>,
+        >(
+            verifier_params,
+            verifying_key,
+            strategy,
+            &[instances],
+            &mut verifier_transcript,
+        )
+        .expect("failed to verify bench circuit");
+    }
+
+    let mock_prover = MockProver::<Fr>::run(degree, &circuit, instance.clone()).unwrap();
```

Comment:
I just saw that the integration tests from the workflow took 56 minutes, because the real prover tests were enabled.  We should split them and not run them on every PR.

Reply:
Using macro is a good idea. With macro, instead of using `mock: true`, maybe can further separate into `test_with_mock_prover`/`test_with_real_prover` for codee clearness